Fix native cache to support non-global IDs

ahmadbaraka opened the pull request:
Changes in this PR address issue #7 to fix the native code to match the JS and support non-global IDs caching

radex commented:
Nice work @ahmadbaraka! 

There are some issues:
- See CI output: https://travis-ci.com/Nozbe/WatermelonDB/jobs/162152637 — Swift doesn't compile and there are Flow errors
- It would be good to have at least a basic test verifying that both adapter.query and adapter.find actually behave correctly. This is the file: https://github.com/Nozbe/WatermelonDB/blob/master/src/adapters/__tests__/commonTests.js

ahmadbaraka commented:
@radex Thanks for quick feedback.
Some changes were done by mistake on the `dev` version in my app instead of being done in the source repo, so they were missed in the commit.

Currently waiting for CI to complete to verify the fixes.

----

> It would be good to have at least a basic test verifying that both adapter.query and adapter.find actually behave correctly. This is the file: https://github.com/Nozbe/WatermelonDB/blob/master/src/adapters/__tests__/commonTests.js

Regarding the tests, I see that file has common tests used for the adapters, but the changes here include implementation details of the SQLite adapter. I see the current tests also verify that `query` and `find` would return either `id` if the model is already cached, or `model._raw` if it's not cached.
Could you point out which test cases we can add here?

radex commented:
> Regarding the tests, I see that file has common tests used for the adapters, but the changes here include implementation details of the SQLite adapter. I see the current tests also verify that query and find would return either id if the model is already cached, or model._raw if it's not cached.
> Could you point out which test cases we can add here?

Right, don't test the implementation details, test the behavior. I'd add a test for both query and find, checking that if I query/find "xxxx" in table A, which is already cached, it will return "xxxx", but if I do the same with table B (where it's not cached), it still returns the whole record.

PS. You only made the change in native implementation, not in LokiJS/web implementation. So you'll have to either also add the LokiJS implementation (should be easy), or, you can just wrap the test in is-native check (in which case, please mark in changelog that this is PARTIAL support for non-global IDs - native only)

ahmadbaraka commented:
I've pushed additional changes to address your comment:
1. Added the LokiJS implementation for non-global IDs
1. Implemented tests for the new changes for `find` and `query` as suggested

rozPierog commented on native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt:
```diff
@@ -129,29 +129,29 @@ class DatabaseDriver(context: Context, dbName: String) {
-        newIds.forEach(this::markAsCached)
-        removedIds.forEach { cachedRecords.remove(it) }
+        newIds.forEach { markAsCached(it.first, it.second) }
+        removedIds.forEach { removeFromCache(it.first, it.second) }
```
```suggestion
        removedIds.forEach { removeFromCache(table = it.first, id = it.second) }
```

rozPierog commented on native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt:
```diff
@@ -163,12 +163,19 @@ class DatabaseDriver(context: Context, dbName: String) {
+        var cache = cachedRecords[table]
+        if (cache == null) {
+            cache = mutableListOf<RecordID>()
+            cachedRecords[table] = cache!!
```
`mutableListOf<RecordID>()` is creating non-nullable type of object so double bang is unnecessary

rozPierog commented on native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt:
```diff
@@ -163,12 +163,19 @@ class DatabaseDriver(context: Context, dbName: String) {
+            cache = mutableListOf<RecordID>()
+            cachedRecords[table] = cache!!
+        }
+        cache!!.add(id)
```
Same here

rozPierog commented on native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt:
```diff
@@ -163,12 +163,19 @@ class DatabaseDriver(context: Context, dbName: String) {
-        cachedRecords.add(id)
+        var cache = cachedRecords[table]
+        if (cache == null) {
+            cache = mutableListOf<RecordID>()
```
Why do you state Type of objects in list if you are leaving it empty?

rozPierog commented on native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt:
```diff
@@ -129,29 +129,29 @@ class DatabaseDriver(context: Context, dbName: String) {
         }
-        newIds.forEach(this::markAsCached)
-        removedIds.forEach { cachedRecords.remove(it) }
+        newIds.forEach { markAsCached(it.first, it.second) }
```
```suggestion
        newIds.forEach { markAsCached(table = it.first, id = it.second) }
```

rozPierog commented on native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt:
```diff
@@ -163,12 +163,19 @@ class DatabaseDriver(context: Context, dbName: String) {
-    private fun isCached(id: RecordID): Boolean = cachedRecords.contains(id)
+    private fun isCached(table: TableName, id: RecordID): Boolean = cachedRecords[table]?.contains(id) ?: false
+
+    private fun removeFromCache(table: TableName, id: RecordID) { cachedRecords[table]?.remove(id) }
```
Why not:
```suggestion
    private fun removeFromCache(table: TableName, id: RecordID) = cachedRecords[table]?.remove(id)
```

radex commented on native/ios/WatermelonDB/DatabaseDriver.swift:
```diff
@@ -163,14 +163,23 @@ class DatabaseDriver {
+            cachedSet = []
+        }
+        cachedSet!.insert(id)
+        cachedRecords[table] = cachedSet
```
Let's do this without force unwrapping. How about:

```
var cachedSet = cachedRecords[table] ?? []
cachedSet.insert(id)
cachedRecords[table] = cachedSet
```

radex commented on src/adapters/__tests__/commonTests.js:
```diff
@@ -112,6 +114,72 @@ export default () => [
+      expectSortedEqual(await adapter.query(taskQuery()), [s1._raw, 'id2'])
+
+      // returns raw if not cached for a different table
+      expectSortedEqual(await adapter.query(projectQuery()), [p1._raw, p2._raw])
```
can you check that the same call the second time will return ids? you're only testing caching after insert works, not that caching after query works

radex commented on src/adapters/__tests__/commonTests.js:
```diff
@@ -112,6 +114,72 @@ export default () => [
+      expect(await adapter.find('tasks', 'id2')).toBe('id2')
+
+      // returns raw if not cached for a different table
+      expect(await adapter.find('projects', 'id2')).toEqual(p2._raw)
```
can you check that the same call the second time will return ids? you're only testing caching after insert works, not that caching after find works

radex commented on src/adapters/lokijs/worker/executor.js:
```diff
@@ -56,8 +56,36 @@ export default class LokiExecutor {
+      return false
+    }
+
+    return cachedSet.has(id)
```
i think we can nicely inline to:

```
return cachedSet ? cachedSet.has(id) : false
```

radex commented on src/adapters/lokijs/worker/executor.js:
```diff
@@ -56,8 +56,36 @@ export default class LokiExecutor {
+      return
+    }
+
+    cachedSet.delete(id)
```
instead of early return, do the delete inside if

radex commented on src/adapters/lokijs/worker/executor.js:
```diff
@@ -56,8 +56,36 @@ export default class LokiExecutor {
+      this.cachedRecords.set(table, cachedSet)
+    }
+
+    cachedSet.add(id)
```
let's do it without local mutability:

cachedSet = xxxx

if ( cachedSet) {
cachedSet.add(x)
} else {
cachedRecords.set(table, new Set(x))

ahmadbaraka commented:
> @rozPierog take a look please ;) — could we avoid mutability in the Kotlin code similar to swift/js?

Fixed

rozPierog commented:
@ahmadbaraka can you fix conflict in `CHANGELOG`? Thanks!

ahmadbaraka commented:
@rozPierog I've resolved the conflict and done a rebase

radex commented:
Cool! I'll merge as soon as the CI status turns green!

@ahmadbaraka Do y'all want some Watermelon stickers? If you do, send me an email and I'll ship you some!

ahmadbaraka commented:
@radex I've sent you an email.

Is the failed build caused somehow by the changes in this PR?

radex commented:
Merging. Thanks @ahmadbaraka — nice work!

Kabangi commented:
When is this expected to be available on NPM?

radex commented:
@Kabangi after New Year — I don't want to break something while I'm away from work. You can just use @ahmadbaraka's fork for now — you'll find the instructions to build production package in Contributing.md
